dotnet/src/Agents/Framework/Internal/BroadcastQueue.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using ChannelQueue = System.Collections.Concurrent.ConcurrentQueue<System.Collections.Generic.IList<Microsoft.SemanticKernel.ChatMessageContent>>;
+
+namespace Microsoft.SemanticKernel.Agents.Internal;
+
+/// <summary>
+/// Utility class used by <see cref="AgentNexus"/> to manage the broadcast of
+/// conversation messages via the <see cref="AgentChannel"/>.
+/// (<see cref="AgentChannel.ReceiveAsync(IEnumerable{ChatMessageContent}, System.Threading.CancellationToken)"/>.)
+/// </summary>
+/// <remarks>
+/// Maintains a set of channel specific queues.
+/// </remarks>
+internal sealed class BroadcastQueue
+{
+    private int _isActive;
+    private readonly Dictionary<string, ChannelQueue> _queues = new();
+    private readonly Dictionary<string, Task> _tasks = new();
+    private readonly Dictionary<string, Exception> _failures = new();
+    private readonly object _queueLock = new(); // Synchronize access to _isActive, _queue and _tasks.
+
+    /// <summary>
+    /// Defines the yield duration when blocking for a channel-queue.
+    /// to drain.
+    /// </summary>
+    public TimeSpan BlockDuration { get; set; } = TimeSpan.FromSeconds(1);
+
+    /// <summary>
+    /// Enqueue a set of messages for a given channel.
+    /// </summary>
+    /// <param name="channels">The target channels for which to broadcast.</param>
+    /// <param name="messages">The messages being broadcast.</param>
+    public void Enqueue(IEnumerable<ChannelReference> channels, IList<ChatMessageContent> messages)
+    {
+        lock (this._queueLock)
+        {
+            foreach (var channel in channels)
+            {
+                var queue = this.GetQueue(channel);
+                queue.Enqueue(messages);
+
+                if (!this._tasks.ContainsKey(channel.Hash))
+                {
+                    this._tasks.Add(channel.Hash, this.ReceiveAsync(channel, queue));
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Blocks until a channel-queue is not in a receive state.
+    /// </summary>
+    /// <param name="channel">A <see cref="ChannelReference"/> structure.</param>
+    /// <returns>false when channel is no longer receiving.</returns>
+    /// <throws>
+    /// When channel is out of sync.
+    /// </throws>
+    public async Task EnsureSynchronizedAsync(ChannelReference channel)
+    {
+        ChannelQueue queue;
+
+        lock (this._queueLock)
+        {
+            if (!this._queues.TryGetValue(channel.Hash, out queue))
+            {
+                return;
+            }
+        }
+
+        while (!queue.IsEmpty) // ChannelQueue is ConcurrentQueue, no need for _queueLock
+        {
+            lock (this._queueLock)
+            {
+                // Activate non-empty queue
```

Comment:
Could you add the reason for the non-empty queue activation to the comment?

Reply:
Really glad you've gotten to this.  I'd really like to spend some cycles with you on evaluating.  At this point, I've significantly invested in functional testing and improvements in the next PR...as introducing AgentChat provided additional insight / experience.  

TLDR - A lot of what is here is already invalidated / evolved / fixed.  On merging this PR we can pivot to that one to continue engagement.

The updated version is really worth digging into.  The version in this PR is fine enough for a framework placeholder with tests.

Reply:
Scheduled an hour on April 8th for us to dig into this.